File: Java/DeepspaceJava/deepspace/src/deepspace/SpaceCity.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author paula
 */
public class SpaceCity extends SpaceStation{
    private ArrayList<SpaceStation> collaborators = new ArrayList<>();
    private SpaceStation base;
    
    SpaceCity (SpaceStation bas, ArrayList<SpaceStation> rest){
        super(bas);
        
        base = bas;
        collaborators = rest;
    }
    
    public ArrayList<SpaceStation> getCollaborators (){
        return collaborators;
    }
    
    @Override
    public float fire (){
        int suma = 0;
        
        for (int i=0; i<collaborators.size(); i++)
            suma += collaborators.get(i).fire();
        
        return suma + super.fire();
    }
    
    @Override
    public SpaceCityToUI getUIversion (){
        return new SpaceCityToUI(this);
    }
    
    @Override
    public float protection (){
        int suma = 0;
        
        for (int i=0; i<collaborators.size(); i++)
            suma += collaborators.get(i).protection();
        
        return suma + super.protection();
    }
    
    @Override
    public Transformation setLoot (Loot loot){
        super.setLoot(loot);
        
        return Transformation.NOTRANSFORM;
    }
    
}

File: Ruby/DeepSpaceRuby/lib/SpaceCity.rb
#encoding:utf-8

require_relative 'SpaceCityToUI'

module Deepspace
  class SpaceCity < SpaceStation
    attr_reader :collaborators

    def initialize (b, rest) #rest : SpaceSation[]
      super(b.name, b.ammoPower, b.fuelUnits, b.shieldPower, b.nMedals, b.shieldBoosters, b.hangar, b.weapons, b.pendingDamage)
      @base = b
      @collaborators = rest
    end

    def getUIversion
      return SpaceCityToUI.new(self)
    end

    def fire
      suma = 0

      @collaborators.each{ |c|
        suma += c.fire
      }

      return suma + super
    end

    def protection
      suma = 0

      @collaborators.each{ |c|
        suma += c.protection
      }

      return suma + super
    end

    def setLoot (loot)
      super

      return Transformation::NOTRANSFORM
    end
  end
end
